File: TypeScript/numIslands.ts
const dfs = (grid: string[][], i: number, j: number) => {
  if (i < 0 || i >= grid.length || j < 0 || j >= grid[i].length || grid[i][j] !== '1') {
    return 0;
  }
  grid[i][j] = '2';

  dfs(grid, i+1, j);
  dfs(grid, i, j+1);
  dfs(grid, i-1, j);
  dfs(grid, i, j-1);

  return 1;
};

const numIslands = (grid: string[][]): number => {
  let islands = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (grid[i][j] === '1') {
        islands += dfs(grid, i, j); 
      }
    }
  }
  return islands;
}


// const grid1 = [
//   ["1", "1", "1", "1", "0"],
//   ["1", "1", "0", "1", "0"],
//   ["1", "1", "0", "0", "0"],
//   ["0", "0", "0", "0", "0"]
// ]

// const grid2 = [
//   ["1", "1", "0", "0", "0"],
//   ["1", "1", "0", "0", "0"],
//   ["0", "0", "1", "0", "0"],
//   ["0", "0", "0", "1", "1"]
// ]

// const grid3 = [
//   ["1", "0", "1", "1", "0"],
//   ["1", "0", "0", "0", "1"],
//   ["0", "0", "1", "0", "0"],
//   ["1", "1", "0", "1", "1"]
// ]


// const result1 = numIslands(grid1); // 1
// const result2 = numIslands(grid2); // 3
// const result3 = numIslands(grid3); // 6

// console.log('result1: ', result1);
// console.log('result2: ', result2);
// console.log('result3: ', result3);

File: Python/num_islands.py
# Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

# An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
# You may assume all four edges of the grid are all surrounded by water.

def dfs(grid, i, j):
    if (i < 0 or
        i >= len(grid) or
        j < 0 or
        j >= len(grid[i]) or
            grid[i][j] != '1'):
        return 0

    grid[i][j] = '2'

    dfs(grid, i + 1, j)
    dfs(grid, i, j + 1)
    dfs(grid, i - 1, j)
    dfs(grid, i, j - 1)

    return 1


def num_islands(grid):
    islands = 0

    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] == '1':
                islands += dfs(grid, i, j)

    return islands


# Test cases

# grid1 = [
#     ["1", "1", "1", "1", "0"],
#     ["1", "1", "0", "1", "0"],
#     ["1", "1", "0", "0", "0"],
#     ["0", "0", "0", "0", "0"]
# ]

# grid2 = [
#     ["1", "1", "0", "0", "0"],
#     ["1", "1", "0", "0", "0"],
#     ["0", "0", "1", "0", "0"],
#     ["0", "0", "0", "1", "1"]
# ]

# grid3 = [
#     ["1", "0", "1", "1", "0"],
#     ["1", "0", "0", "0", "1"],
#     ["0", "0", "1", "0", "0"],
#     ["1", "1", "0", "1", "1"]
# ]

# grid1_result = num_islands(grid1)  # 1
# grid2_result = num_islands(grid2)  # 3
# grid3_result = num_islands(grid3)  # 6

# print('grid1_result: ', grid1_result)
# print('grid2_result: ', grid2_result)
# print('grid3_result: ', grid3_result)
